lib/legacy/legacy-data-mixin.js

```diff
@@ -0,0 +1,150 @@
+ * computing functions that need undefined checks to work without
+ * the mixin in Polymer 2.
+ *
+ * @mixinFunction
+ * @polymer
+ * @summary Mixin to selectively add back Polymer 1.x's `undefined` rules
+ * governing when observers & computing functions run.
+ */
+export const LegacyDataMixin = dedupingMixin(superClass => {
+
+  /**
+   * @constructor
+   * @extends {superClass}
+   * @unrestricted
+   * @private   */
+  class LegacyDataMixin extends superClass {
+    /**
+     * Overrides `Polymer.PropertyEffects` to add `undefined` argument
+     * checking to match Polymer 1.x style rules
+     * 
+     * @param {!Array<!MethodArg>} args Array of argument metadata
+     * @param {string} path Property/path name that triggered the method effect
+     * @param {Object} props Bag of current property changes
+     * @return {Array<*>} Array of argument values
+     * @private
+     */
+    _marshalArgs(args, path, props) {
+      const vals = super._marshalArgs(args, path, props);
+      // Per legacy data rules, single-property observers (whether in `properties`
+      // and in `observers`) are called regardless of whether their argument is
+      // undefined or not. Multi-property observers must have all arguments defined
+      if (this._legacyUndefinedCheck && vals.length > 1) {
+        for (let i=0; i<vals.length; i++) {
+          if (vals[i] === undefined) {
+            // Break out of effect's control flow; will be caught in
+            // wrapped property effect function below
+            const name = args[i].name;
+            throw new UndefinedArgumentError(`Argument '${name}' is undefined. Ensure it has an undefined check.`, name);
+          }
+        }
+      }
+      return vals;
+    }
+
+    /**
+     * Overrides `Polyer.PropertyEffects` to wrap effect functions to
+     * catch `UndefinedArgumentError`s and warn.
+     * 
+     * @param {string} property Property that should trigger the effect
+     * @param {string} type Effect type, from this.PROPERTY_EFFECT_TYPES
+     * @param {Object=} effect Effect metadata object
+     * @return {void}
+     * @protected
+     */
+    _addPropertyEffect(property, type, effect) {
+      return super._addPropertyEffect(property, type,
+        wrapEffect(effect, effect && effect.info && effect.info.methodName));
+    }
+
+    /**
+     * Overrides `Polyer.PropertyEffects` to wrap effect functions to
+     * catch `UndefinedArgumentError`s and warn.
+     *
+     * @param {Object} templateInfo Template metadata to add effect to
+     * @param {string} prop Property that should trigger the effect
+     * @param {Object=} effect Effect metadata object
+     * @return {void}
+     * @protected
+     */
+    static _addTemplatePropertyEffect(templateInfo, prop, effect) {
+      return super._addTemplatePropertyEffect(templateInfo, prop, wrapEffect(effect));
+    }
+
+  }
+
+  return LegacyDataMixin;
+
+});
+
+Polymer.Class = (info, mixin) => Class(info, 
```

Comment:
Why not just put `LegacyDataMixin` on top of the user class?

Reply:
Ok, the answer is no because the mixin customizes how properties are created on the element prototype. Since a subclass only creates properties that it defines, it will not modify properties created by its superclass.